Feature: linkcheck-action

HunterGerlach opened the pull request:
- Add support for local build and linkchecking of docs
- Add GitHub Action to automatically check links before prior to publication
- Minor reformat of Contributing readme
- Fix existing broken links

HunterGerlach committed:
minor grammar fixes for memory docs

HunterGerlach committed:
Merge remote-tracking branch 'upstream/master'

HunterGerlach committed:
Merge remote-tracking branch 'upstream/master'

HunterGerlach committed:
Merge remote-tracking branch 'upstream/master'

HunterGerlach committed:
Merge commit '985496f4be81e2e096a9da5b0a3791033175add7'

hwchase17 commented on .github/workflows/linkcheck.yml:
```diff
@@ -0,0 +1,36 @@
+    runs-on: ubuntu-latest
+    strategy:
+      matrix:
+        python-version:
```
probably only needs to run on one python version

HunterGerlach replied:
Good call.

hwchase17 reviewed (commented):
yes! love this - thank you so much!!!!

HunterGerlach commented:
@hwchase17 I am running into an issue with the link re-writes. Since they are rewritten from `.rst`, `.md`, and `.ipynb`, they are unable to be found by `linkcheck`. I put a brute force example in the `conf.py` of how we _could_ solve this if necessary, but I would much rather have a clean method. Unfortunately, I'm not too well-versed on Sphinx to know if there's a "right" way to do this or not. Any ideas you have would be appreciated.

HunterGerlach committed:
Merge remote-tracking branch 'origin/master'

HunterGerlach committed:
Add enhanced support for building docs locally

HunterGerlach committed:
Add linkcheck GitHub Action

HunterGerlach committed:
Additional broken link fix

HunterGerlach committed:
Partial broken link fix

HunterGerlach committed:
WIP to ignore erroneous links

HunterGerlach committed:
Reduce number of python versions tested.

HunterGerlach commented:
I don't mind doing it to free you up for other more important tasks if you just want to do a few so I can see an example or two (e.g. `.md` and `.ipynb`).

hwchase17 commented:
> I don't mind doing it to free you up for other more important tasks if you just want to do a few so I can see an example or two (e.g. `.md` and `.ipynb`).

i *think* (not 100%, need to spend some time to be sure) that ":doc:`foo <utils/how_to_guides>`:" should work?

HunterGerlach commented:
We'll see...

HunterGerlach committed:
Fix outstanding broken links

HunterGerlach committed:
Adjust to using new/specific linkchecker tool

HunterGerlach committed:
Add docs dependencies for build/linkcheck

HunterGerlach committed:
Update action to run via poetry

HunterGerlach commented:
> > I don't mind doing it to free you up for other more important tasks if you just want to do a few so I can see an example or two (e.g. `.md` and `.ipynb`).
> 
> i _think_ (not 100%, need to spend some time to be sure) that ":doc:`foo <utils/how_to_guides>`:" should work?

I was unable to get it to work with `make linkcheck` - I think it's because it's checking the pre-compiled links _before_ their extensions change and it gets confused.

I switched to using a dedicated linkchecker instead.

I have removed all broken links, so running this from now on should prevent any broken links from getting into `master`.

hwchase17 commented:
@HunterGerlach .... i dont think it is an overstatement to call you a hero for this

im gonna merge this into a dev branch and build a copy of the docs from that (just to check that the build-the-docs action works as expected with this)
